adlfs/transfer.py

```diff
@@ -267,110 +310,101 @@ def progress(self):
-                future = dic['merge']
-                if not future.done():
-                    continue
-                if future.cancelled():
-                    self._fstates[(src, dst)] = 'cancelled'
-                elif future.exception():
-                    self._fstates[(src, dst)] = 'errored'
+    def _update(self, future):
+        if future in self._cfutures:
+            obj = self._cfutures[future]
+            parent = self._chunks[obj]['parent']
+            cstates = self._files[parent]['cstates']
+
+            if future.cancelled():
+                cstates[obj] = 'cancelled'
+            elif future.exception():
+                cstates[obj] = 'errored'
+            else:
+                self._chunks[obj]['actual'] = future.result()
+                cstates[obj] = 'finished'
+
+            if cstates.contains_all('finished'):
+                logger.debug("Chunks transferred")
+                src, dst = parent
+                if self._merge and len(cstates.objects) > 1:
+                    logger.debug("Merging file: %s", self._fstates[parent])
+                    self._fstates[parent] = 'merging'
+                    merge_future = self._submit(
+                        self._merge, self._adlfs, dst,
+                        [o[0] for o in cstates.objects])
+                    self._ffutures[merge_future] = parent
                 else:
-                    dic['stop'] = time.time()
-                    self._fstates[(src, dst)] = 'finished'
-                    self._status(src, dst, dic['nbytes'], dic['start'], dic['stop'])
+                    self._fstates[parent] = 'finished'
+                    self._files[parent]['stop'] = time.time()
+                    self._status(src, dst)
+            elif cstates.contains_none('running'):
+                logger.debug("Transfer failed: %s", cstates)
+                self._fstates[parent] = 'errored'
+        elif future in self._ffutures:
+            src, dst = self._ffutures[future]
+
+            if future.cancelled():
+                self._fstates[(src, dst)] = 'cancelled'
+            elif future.exception():
+                self._fstates[(src, dst)] = 'errored'
+            else:
+                result = future.result()
+                self._fstates[(src, dst)] = 'finished'
+                self._files[(src, dst)]['stop'] = time.time()
+                self._status(src, dst)
         self.save()
 
-    def run(self, nthreads=None, monitor=True, before_scatter=None):
+    def run(self, nthreads=None, monitor=True, before_start=None):
         self._nthreads = nthreads or self._nthreads
         for src, dst in self._files:
-            self._files[(src, dst)]['start'] = time.time()
-            self._fstates[(src, dst)] = 'transferring'
-            if before_scatter:
-                before_scatter(self._adlfs, src, dst)
-            self._scatter(src, dst, self._transfer)
+            if before_start:
+                before_start(self._adlfs, src, dst)
```

Comment:
This is a local function, not `adlfs.touch`. I pass both parameters to be generic. These two lines were included when creating the generic client, but they may be removed at some point.